greenlight/harness/testcase.py

```diff
@@ -19,10 +19,17 @@ def setUp(self, *args, **kwargs):
         MarionetteTestCase.setUp(self, *args, **kwargs)
         Puppeteer.set_marionette(self, self.marionette)
 
+        self._start_handle_count = len(self.marionette.window_handles)
         self.marionette.set_context('chrome')
         self.browser = self.windows.current
 
     def tearDown(self, *args, **kwargs):
-        MarionetteTestCase.tearDown(self, *args, **kwargs)
+        win_count = len(self.marionette.window_handles)
+        self.assertEqual(win_count, self._start_handle_count,
+                         "A test must not leak window handles. "
+                         "This test started the browser with %s open "
+                         "top level browsing contexts, but ended with %s." %
+                         (self._start_handle_count, win_count))
 
         self.prefs.restore_all_prefs()
+        MarionetteTestCase.tearDown(self, *args, **kwargs)
```

Comment:
We talked about this on irc, and agreed that while there's more to be done in terms of a robust tearDown of each test case, this assertion is worth taking so long as it runs after all other cleanup to in case of failure it doesn't make matters worse for further tests.